src/backend/utils/datumstream/datumstreamblock.c

```diff
@@ -1642,7 +1642,8 @@ DatumStreamBlockWrite_PutOrig(
 			p = DatumGetPointer(d);
 			wsz = sz;
 		}
-		else if (value_type_could_short(DatumGetPointer(d), dsw->typeInfo->typid))
+		else if (dsw->typeInfo->storage != 'p' &&
+				 value_type_could_short(DatumGetPointer(d), dsw->typeInfo->typid))
```

Comment:
Why not add this additional check inside the `value_type_could_short` function?
```suggestion
		else if (value_type_could_short(DatumGetPointer(d), dsw->typeInfo->typid, dsw->typeInfo->storage))
```

Reply:
Fixed